sourcegraph/sourcegraph
# git-debuggability: paginate recorded commands

[Walkthrough](https://www.loom.com/share/8278ed0143af4f3d8734a69dafce91b8)

This PR introduces a slice connection resolver that can be used to paginate slices. I added this under the `graphqlutil` package because it fulfills an interface similar to that of the `ConnectionResolver`.

```graphql
type SliceConnectionResolver[T, any] interface {
	Nodes(ctx context.Context) []T
	TotalCount(ctx context.Context) int32
	PageInfo(ctx context.Context) *PageInfo
}
```

This ensures we can adequately paginate the `RecordedCommands` field on the `RepositoryResolver`.

This is particularly useful because we get a slice of byte slice from redis and need to transform that into `RecordedCommandResolver`.

## Test plan



* Navigate to the API console
* Run a GraphQL query to fetch the recorded commands for any repository of your choosing.
```graphql
query {
  repository(name:  <REPOSITORY_NAME>) {
    id
    name
    mirrorInfo {
      shard
    }
    recordedCommands(limit: 100) {
      totalCount
      pageInfo {
        hasNextPage
      }
      nodes {
        dir
        command
        duration
        start
      }
    }
  }
}
```

* You should be able to retrieve the nodes, pageInfo and totalCount of the recorded commands.
